docs/ReferenceGuides/UnderscoredAttributes.md

```diff
@@ -0,0 +1,131 @@
+# Underscored Attributes Reference
+
+**WARNING:** This information is provided primarily for compiler and standard
+library developers. Usage of these attributes outside of the Swift monorepo
+is strongly discouraged.
+
+The Swift reference has a chapter discussing [stable attributes][Attributes].
+This document is intended to serve as a counter-part describing underscored
+attributes, whose semantics are subject to change and most likely need to
+go through the Swift evolution process before being stabilized.
+
+The attributes are organized in alphabetical order.
+
+## `@_alignment`
+
+## `@_borrowed`
+
+## `@_cdecl`
+
+Similar to `@_silgen_name` but uses the C calling convention.
+This attribute doesn't have very well-defined semantics.
+
+For potential ideas on stabilization, see
+[Formalizing `@cdecl`](https://forums.swift.org/t/formalizing-cdecl/40677).
+
+## `@_dynamicReplacement(for:)`
+
+Marks a function as the dynamic replacement for another `dynamic` function.
+This is similar to method swizzling in other languages such as Objective-C,
+except that the replacement happens at program start (or loading a shared
+library), instead of at an arbitrary point in time.
+
+For more details, see [Dynamic method replacement](https://forums.swift.org/t/dynamic-method-replacement/16619).
+
+## `@_effects`
+
+## `@_exported`
+
+Use to mark an imported module to re-export all its declarations.
+This attribute is most commonly used by overlays.
+
+```
+// module M
+func f() {}
+
+// module N
+@_exported import M
+
+// module P
+import N
+let () = N.f() // OK
+```
+
+## `@_fixed_layout`
+
+## `@_hasInitialValue`
+
+## `@_hasMissingDesignatedInitializers`
+
+## `@_hasStorage`
+
+## `@_implementationOnly`
+
+Used to mark an imported module as an implementation detail.
+This prevents types from that module being exposed in API
+(types of public functions, constraints in public extension etc.)
+and ABI (usage in `@inlinable` code).
+
+## `@_implements`
+
+## `@_inheritsConvenienceInitializers`
+
+## `@_nonEphemeral`
```

Comment:
````
## `@_nonEphemeral`

Marks a function parameter that cannot accept a temporary pointer produced from
an inout-to-pointer, array-to-pointer, or string-to-pointer conversion. Such a
parameter may only accept a pointer that is guaranteed to outlive the duration
of the function call.

Attempting to pass a temporary pointer to an `@_nonEphemeral` parameter will
produce a warning:

```swift
func foo(@_nonEphemeral _ ptr: UnsafePointer<Int>) {}

func bar(_ ptr: UnsafePointer<Int>) {
  var x = 0

  // warning: inout expression creates a temporary pointer, but argument #1
  // should be a pointer that outlives the call to 'foo'
  foo(&x)

  // warning: passing '[Int]' to parameter, but argument #1 should be a pointer
  // that outlives the call to 'foo'
  foo([1, 2, 3])

  foo(ptr) // Okay
}
```

This attribute is primarily used within the standard library on the various
`UnsafePointer` initializers to warn users about the undefined behavior caused
by using a temporary pointer conversion as an argument:

```swift
func baz() {
  var x = 0

  // warning: Initialization of 'UnsafePointer<Int>' results in a dangling pointer
  let ptr = UnsafePointer(&x)

  // warning: Initialization of 'UnsafePointer<Int>' results in a dangling pointer
  let ptr2 = UnsafePointer([1, 2, 3])
}
```

This is because the temporary pointer conversion produces a pointer that is only
guaranteed to be valid for the duration of the call to the initializer, but will
be invalid once the initialized `UnsafePointer` is returned to the caller,
thereby leaving a dangling pointer.

Note we do guarantee that inout-to-pointer conversions on static and global
variables produce non-ephemeral pointers, as long as they are stored and have no
observers:

```swift
var global = 0

struct S {
  static var staticVar = 0
}

func baz() {
  let ptr = UnsafePointer(&global) // Okay
  let ptr2 = UnsafePointer(&S.staticVar) // Also okay
}
```

Additionally, if they are of a tuple or struct type, their stored members
without observers may also be passed inout as non-ephemeral pointers.

For more information, see [the educational note on temporary pointer usage](/userdocs/diagnostics/temporary-pointers.md).
````